ntoskrnl/io/iomgr/deviface.c

```diff
@@ -304,7 +305,177 @@ IoGetDeviceInterfaceAlias(IN PUNICODE_STRING SymbolicLinkName,
                           IN CONST GUID *AliasInterfaceClassGuid,
                           OUT PUNICODE_STRING AliasSymbolicLinkName)
 {
-    return STATUS_NOT_IMPLEMENTED;
+    UNICODE_STRING BaseKeyName = RTL_CONSTANT_STRING(L"\\Registry\\Machine\\System\\CurrentControlSet\\Control\\Class\\");
+    UNICODE_STRING AliasKeyName = RTL_CONSTANT_STRING(L"\\Registry\\Machine\\System\\CurrentControlSet\\Services\\");
+    UNICODE_STRING bipName, InLinkName;
+    UNICODE_STRING GuidString;
+    HANDLE InterfaceKey = NULL;
+    PKEY_FULL_INFORMATION fip = NULL;
+    PKEY_BASIC_INFORMATION bip = NULL;
+    PWCHAR pdest;
+    ULONG i, Size;
+    LONG ret;
+    NTSTATUS Status;
+
+    PAGED_CODE();
+
+    /* Convert alias GUID to a string */
+    Status = RtlStringFromGUID(AliasInterfaceClassGuid, &GuidString);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("RtlStringFromGUID() failed with status 0x%08lx\n", Status);
+        goto cleanup;
+    }
+
+    BaseKeyName.Length = 0;
+    BaseKeyName.MaximumLength += GuidString.Length;
+    BaseKeyName.Buffer = ExAllocatePool(PagedPool, BaseKeyName.MaximumLength);
```

Comment:
Please use `ExAllocatePoolWithTag` whenever possible. The tag to use for such pool allocation operations is TAG_IO (see other examples in this code).

Reply:
OK. And I seem should use `ExFreePoolWithTag` for each appropriate `ExAllocatePoolWithTag` call? Or it is not necessarily? :thinking:

Reply:
yes it is.

Reply:
Thanks.

Reply:
Done. :ok_hand: